core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -13,331 +13,321 @@
  * License for the specific language governing permissions and limitations
  * under the License.
  */
+
 package com.linecorp.armeria.client.retry;
 
+import static com.google.common.base.MoreObjects.firstNonNull;
 import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.base.Preconditions.checkState;
-import static java.util.Objects.requireNonNull;
+import static com.linecorp.armeria.internal.ClientUtil.executeWithFallback;
 
-import java.util.concurrent.TimeUnit;
-import java.util.function.Consumer;
+import java.time.Duration;
+import java.util.Date;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import java.util.function.Function;
 
 import javax.annotation.Nullable;
 
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.linecorp.armeria.client.Client;
-import com.linecorp.armeria.client.ClientFactory;
 import com.linecorp.armeria.client.ClientRequestContext;
-import com.linecorp.armeria.client.Endpoint;
-import com.linecorp.armeria.client.SimpleDecoratingClient;
-import com.linecorp.armeria.client.endpoint.EndpointSelector;
+import com.linecorp.armeria.client.HttpClient;
+import com.linecorp.armeria.client.ResponseTimeoutException;
+import com.linecorp.armeria.common.FilteredHttpResponse;
+import com.linecorp.armeria.common.HttpData;
 import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
 import com.linecorp.armeria.common.HttpRequest;
-import com.linecorp.armeria.common.Request;
-import com.linecorp.armeria.common.RequestId;
-import com.linecorp.armeria.common.Response;
-import com.linecorp.armeria.common.RpcRequest;
+import com.linecorp.armeria.common.HttpRequestDuplicator;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpResponseDuplicator;
+import com.linecorp.armeria.common.RequestHeadersBuilder;
+import com.linecorp.armeria.common.logging.RequestLogAvailability;
+import com.linecorp.armeria.common.stream.AbortedStreamException;
 
-import io.netty.util.AsciiString;
-import io.netty.util.AttributeKey;
-import io.netty.util.concurrent.ScheduledFuture;
+import io.netty.handler.codec.DateFormatter;
 
 /**
- * A {@link Client} decorator that handles failures of remote invocation and retries requests.
- *
- * @param <I> the {@link Request} type
- * @param <O> the {@link Response} type
+ * An {@link HttpClient} decorator that handles failures of an invocation and retries HTTP requests.
  */
-public abstract class RetryingClient<I extends Request, O extends Response>
-        extends SimpleDecoratingClient<I, O> {
+public final class RetryingClient extends AbstractRetryingClient<HttpRequest, HttpResponse>
```

Comment:
We had an offline chat and decided to rename it back to `RetryingHttpClient`, so that this pull request focuses on type parameter changes rather than renaming. A follow-up PR that focuses on renaming will follow later (after releasing 0.96).